### PR TITLE
Problem: improve permission checking in a few messages

### PR DESCRIPTION
**Pull request by yihuang**
👮🏻👮🏻👮🏻 !!!! REFERENCE THE PROBLEM YOUR ARE SOLVING IN THE PR TITLE AND DESCRIBE YOUR SOLUTION HERE !!!! DO NOT FORGET !!!! 👮🏻👮🏻👮🏻

# PR Checklist:

- [ ] Have you read the [CONTRIBUTING.md](https://github.com/crypto-org-chain/chain-main/blob/master/CONTRIBUTING.md)?
- [ ] Does your PR follow the [C4 patch requirements](https://rfc.zeromq.org/spec:42/C4/#23-patch-requirements)?
- [ ] Have you rebased your work on top of the latest master? 
- [ ] Have you checked your code compiles? (`make`)
- [ ] Have you included tests for any non-trivial functionality?
- [ ] Have you checked your code passes the unit tests? (`make test`)
- [ ] Have you checked your code formatting is correct? (`go fmt`)
- [ ] Have you checked your basic code style is fine? (`golangci-lint run`)
- [ ] If you added any dependencies, have you checked they do not contain any known vulnerabilities? (`go list -json -m all | nancy sleuth`)
- [ ] If your changes affect the client infrastructure, have you run the integration test?
- [ ] If your changes affect public APIs, does your PR follow the [C4 evolution of public contracts](https://rfc.zeromq.org/spec:42/C4/#26-evolution-of-public-contracts)?
- [ ] If your code changes public APIs, have you incremented the crate version numbers and documented your changes in the [CHANGELOG.md](https://github.com/crypto-org-chain/chain-main/blob/master/CHANGELOG.md)?
- [ ] If you are contributing for the first time, please read the agreement in [CONTRIBUTING.md](https://github.com/crypto-org-chain/chain-main/blob/master/CONTRIBUTING.md) now and add a comment to this pull request stating that your PR is in accordance with the [Developer's Certificate of Origin](https://github.com/crypto-org-chain/chain-main/blob/master/CONTRIBUTING.md#developer-certificate-of-originn).

Thank you for your code, it's appreciated! :)

## Summary by CodeRabbit

- **Security Enhancements**
  - Improved permission checking for message processing to enhance security.

- **Bug Fixes**
  - Permission checks added to `TransferTokens`, `UpdateTokenMapping`, and `TurnBridge` methods to ensure proper authorization.

- **Refactor**
  - Removed unnecessary imports from various files to streamline codebase.
  - Updated `HasPermission` function to handle multiple accounts, improving flexibility in permission checks.
  - Adjusted test cases to align with the updated permission checking logic.

- **Documentation**
  - Updated CHANGELOG to reflect the security and permission enhancements.

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -11,6 +11,7 @@
 - [#1216](https://github.com/crypto-org-chain/cronos/pull/1216) Update ethermint to fix of avoid redundant parse chainID from gensis when start server.
 - [#1230](https://github.com/crypto-org-chain/cronos/pull/1230) Fix mem store in versiondb multistore.
 - [#1233](https://github.com/crypto-org-chain/cronos/pull/1233) Re-emit logs in callback contract.
+- [#]() Improve permission checkings for some messages.
```
Comment (yihuang) [inline CHANGELOG.md:14]:
```suggestion
- [#1256](https://github.com/crypto-org-chain/cronos/pull/1256) Improve permission checkings for some messages.
```

File: x/cronos/keeper/permissions.go [hunk 2]
```diff
@@ -28,13 +28,21 @@ func (k Keeper) GetPermissions(ctx sdk.Context, address sdk.AccAddress) uint64 {
 }
 
 // HasPermission check if an account has a specific permission. by default cronos admin has all permissions
-func (k Keeper) HasPermission(ctx sdk.Context, account sdk.AccAddress, permissionsToCheck uint64) bool {
+func (k Keeper) HasPermission(ctx sdk.Context, accounts []sdk.AccAddress, permissionsToCheck uint64) bool {
```
Comment (mmsqe) [inline x/cronos/keeper/permissions.go:31]: might fix permission test as well

Comment (yihuang) [reply x/cronos/keeper/permissions.go:31]: fixed committed

File: x/cronos/keeper/msg_server.go [hunk 3]
```diff
@@ -66,6 +66,12 @@ func (k msgServer) TransferTokens(goCtx context.Context, msg *types.MsgTransferT
 // UpdateTokenMapping implements the grpc method
 func (k msgServer) UpdateTokenMapping(goCtx context.Context, msg *types.MsgUpdateTokenMapping) (*types.MsgUpdateTokenMappingResponse, error) {
 	ctx := sdk.UnwrapSDKContext(goCtx)
+
+	// check permission
+	if !k.Keeper.HasPermission(ctx, msg.GetSigners(), CanChangeTokenMapping) {
+		return nil, sdkerrors.Wrap(sdkerrors.ErrUnauthorized, "msg sender is not authorized")
```
Comment (mmsqe) [inline x/cronos/keeper/msg_server.go:72]: might use errors.Wrap here

Comment (yihuang) [reply x/cronos/keeper/msg_server.go]: fixed

Comment (yihuang) [reply x/cronos/keeper/msg_server.go]: fixed

**Review by mmsqe — APPROVED** [review]

File: x/cronos/keeper/msg_server.go [hunk 4]
```diff
@@ -101,7 +113,7 @@ func (k msgServer) UpdatePermissions(goCtx context.Context, msg *types.MsgUpdate
 	admin := k.Keeper.GetParams(ctx).CronosAdmin
 	// if admin is empty, no sender could be equal to it
 	if admin != msg.From {
-		return nil, errors.Wrap(sdkerrors.ErrInvalidAddress, "msg sender is authorized")
+		return nil, errors.Wrap(sdkerrors.ErrInvalidAddress, "msg sender is not authorized")
```
Comment (mmsqe) [inline x/cronos/keeper/msg_server.go:116]:
```suggestion
		return nil, errors.Wrap(sdkerrors.ErrUnauthorized, "msg sender is not authorized")
```

**Pull request merged**